x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart/MlMigrationFullClusterRestartIT.java

```diff
@@ -0,0 +1,258 @@
+        putStoppedDatafeed.setJsonEntity(Strings.toString(stoppedDfBuilder.build()));
+        client().performRequest(putStoppedDatafeed);
+    }
+
+    private void upgradedClusterTests() throws Exception {
+        // wait for the closed job and datafeed to be migrated
+        waitForMigration(Collections.singletonList(OLD_CLUSTER_CLOSED_JOB_ID),
+                Collections.singletonList(OLD_CLUSTER_STOPPED_DATAFEED_ID),
+                Collections.singletonList(OLD_CLUSTER_OPEN_JOB_ID),
+                Collections.singletonList(OLD_CLUSTER_STARTED_DATAFEED_ID));
+
+        // the job and datafeed left open during upgrade should
+        // be assigned to a node
+        waitForJobToBeAssigned(OLD_CLUSTER_OPEN_JOB_ID);
+        waitForDatafeedToBeAssigned(OLD_CLUSTER_STARTED_DATAFEED_ID);
+
+        // open the migrated job and datafeed
+        Request openJob = new Request("POST", "_xpack/ml/anomaly_detectors/" + OLD_CLUSTER_CLOSED_JOB_ID + "/_open");
+        client().performRequest(openJob);
+        Request startDatafeed = new Request("POST", "_xpack/ml/datafeeds/" + OLD_CLUSTER_STOPPED_DATAFEED_ID + "/_start");
+        client().performRequest(startDatafeed);
+
+        // close the job left open during upgrade
+        Request stopDatafeed = new Request("POST", "_xpack/ml/datafeeds/" + OLD_CLUSTER_STARTED_DATAFEED_ID + "/_stop");
+        client().performRequest(stopDatafeed);
+
+        Request closeJob = new Request("POST", "_xpack/ml/anomaly_detectors/" + OLD_CLUSTER_OPEN_JOB_ID + "/_close");
+        client().performRequest(closeJob);
+
+        // now all jobs should be migrated
+        waitForMigration(Arrays.asList(OLD_CLUSTER_CLOSED_JOB_ID, OLD_CLUSTER_OPEN_JOB_ID),
+                Arrays.asList(OLD_CLUSTER_STOPPED_DATAFEED_ID, OLD_CLUSTER_STARTED_DATAFEED_ID),
+                Collections.emptyList(),
+                Collections.emptyList());
+    }
+
+    @SuppressWarnings("unchecked")
+    private void waitForJobToBeAssigned(String jobId) throws Exception {
+        assertBusy(() -> {
+            try {
+                Request getJobStats = new Request("GET", "_xpack/ml/anomaly_detectors/" + jobId + "/_stats");
+                Response response = client().performRequest(getJobStats);
+
+                Map<String, Object> stats = entityAsMap(response);
+                List<Map<String, Object>> jobStats =
+                        (List<Map<String, Object>>) XContentMapValues.extractValue("jobs", stats);
+
+                assertEquals(jobId, XContentMapValues.extractValue("job_id", jobStats.get(0)));
+                assertEquals("opened", XContentMapValues.extractValue("state", jobStats.get(0)));
+                assertThat((String)XContentMapValues.extractValue("assignment_explanation", jobStats.get(0)), isEmptyOrNullString());
+                assertNotNull(XContentMapValues.extractValue("node", jobStats.get(0)));
+            } catch (IOException e) {
```

Comment:
Why do we need to ignore `IOException`s here?

Reply:
Well the assert busy will retry and eventually timeout, either way the test will fail but I admit the cause will be lost

Reply:
I don't do it elsewhere I'll clean that up

Reply:
I think an exception other than an `AssertionError` will stop the busy loop.  So if an `IOException` can be thrown that _shouldn't_ terminate the busy loop then it needs to be caught here.  But maybe there isn't.

Reply:
Yes assert busy only catches  `AssertionError` and the IOException should fail the test and not be caught here. I'm gonna blame intellij and remove the try...catch